Stories: Add Intro screen

bjtitus opened the pull request:
Adds the Intro Screen for the Stories feature.
Equivalent Android PR: https://github.com/wordpress-mobile/WordPress-Android/pull/13141

## Changes

Some details about the `StoriesIntroViewController` implementation:
- Reuses UI from the What Is New feature.
- Manually constructs its own data source (`StoriesIntroDataSource`) instead of reusing the What Is New announcement data sources since many of the fields were unnecessary for this screen.

Some details about the new cell type, `GridCell`: 
- Consists of several nested Stack Views to construct the side by side example story view. These are outlined in doc comments for `GridCell` and `makeItemStackView`.
- `Item` views (example stories) are left aligned in Regular width size class but centered in a Compact size class.
- `Item` views scale down to a smaller width when the width becomes too small to fit them at full size.

A few other implementation details:
- The `WhatIsNewView` has a new back button to dismiss the screen instead of continuing.
- I've created a new `URLQueryItem+Parameters` file to store commonly used query parameters within namespaces so they are easier to discover and audit.
- A new `URL.appendingQueryItems` makes using the defined query parameters easier.

### Screenshots

<details>
  <summary><strong>iPhone</strong></summary>
<a href="https://user-images.githubusercontent.com/3250/96625045-e0d9dc80-12ca-11eb-9ecf-7f96c42bb0bf.png"><img src="https://user-images.githubusercontent.com/3250/96625045-e0d9dc80-12ca-11eb-9ecf-7f96c42bb0bf.png" width="300"></a> <a href="https://user-images.githubusercontent.com/3250/96625408-665d8c80-12cb-11eb-921a-15763570fdd2.png"><img src="https://user-images.githubusercontent.com/3250/96625408-665d8c80-12cb-11eb-921a-15763570fdd2.png" width="300"></a>

<a href="https://user-images.githubusercontent.com/3250/96625490-7d03e380-12cb-11eb-9c5b-898e1b2da0e3.png"><img src="https://user-images.githubusercontent.com/3250/96625490-7d03e380-12cb-11eb-9c5b-898e1b2da0e3.png" width="500"></a>
</details>

<details>
  <summary><strong>iPad</strong></summary>

<a href="https://user-images.githubusercontent.com/3250/96625987-1e8b3500-12cc-11eb-84af-24e5d002e10c.png"><img src="https://user-images.githubusercontent.com/3250/96625987-1e8b3500-12cc-11eb-84af-24e5d002e10c.png" width="400"></a>  <a href="https://user-images.githubusercontent.com/3250/96626056-38c51300-12cc-11eb-85b4-782935bf7f3f.png"><img src="https://user-images.githubusercontent.com/3250/96626056-38c51300-12cc-11eb-85b4-782935bf7f3f.png" width="400"></a>

<a href="https://user-images.githubusercontent.com/3250/96626147-54c8b480-12cc-11eb-949e-74335d22da1a.png"><img src="https://user-images.githubusercontent.com/3250/96626147-54c8b480-12cc-11eb-949e-74335d22da1a.png" width="400"></a>

</details>

### Testing
- Ensure that the Stories Intro screen shows when tapping "Story post" from the FAB
- Tapping on the image for either example post should show a fullscreen web view which autoplays the story.
- Dismissing the Stories Intro Screen should work using the back button or by swiping down on iOS 14
- Tapping "Create Story Post" should dismiss and show a red screen (can dismiss by swiping down). This is where the Kanvas UI will go.
- After tapping "Create Story Post" tapping "Story post" from the FAB menu should show the red screen instead of the Stories Intro Screen.

#### Logging
- `story_intro_shown` should be logged whenever the Stories Intro Screen is presented.
- `story_intro_dismissed` should be logged when the Stories Intro Screen is dismissed using the Back button or swiping down.
- `story_intro_create_story_button_tapped` should be logged when the "Create Story Post" button is tapped.

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

bjtitus committed:
Add back button to What Is New screen

bjtitus committed:
Add Stories Intro screen

Includes analytics screens + analytics events

bjtitus committed:
Fix centering in landscape mode

bjtitus committed:
Add Stories Intro UserDefaults flag

Toggles when the "Create Story Post" has been pressed

bjtitus committed:
Refactor StoriesIntro into WhatIsNew subclass

bjtitus committed:
Stories Intro refactor + Accessibility improvements

bjtitus committed:
Fix Stories Intro layout, navigation bar & button

- Fix the Stories Intro layout in ultracompact views (iPhone SE and narrowest iPad split screen)
- Fix the sample story buttons not working in < iOS 14
- Add a navigation bar to the sample story web view so it can be dismissed < iOS 14.

bjtitus committed:
Doc cleanup, namespaced URLQueryItems, URL appending

- Adds an appending function to URL+Helpers
- Moves the namespaced URLQueryItems to a new extension.
- Minor doc comment clean up.

bjtitus committed:
Remove GridCell iOS 14 button + UIAction

This is not available in CI yet so the project will fail to compile.

bjtitus committed:
Rename `URL.appendingQueryItems`

To better match current URL method naming (`appendingPathComponents`)

bjtitus committed:
Move StoriesIntroDataSource into its own file

bjtitus commented:
@Gio2018 Would love to hear your thoughts on reusing `WhatIsNew` in this way and whether the `SourceIntroDataSource` makes sense.

aerych commented on WordPress/Classes/Extensions/URL+Helpers.swift:
```diff
@@ -206,3 +206,16 @@ extension URL {
+    func appendingQueryItems(_ newQueryItems: [URLQueryItem]) -> URL {
+        var components = URLComponents(url: self, resolvingAgainstBaseURL: false)
+        var queryItems = components?.queryItems ?? []
+        queryItems.append(contentsOf: newQueryItems)
```
I'm waffling a bit here.  Since this is a general extension I'm thinking about other uses cases. Particularly what is the best behavior for when a parameter already exists.  Having the parameter twice could produce unexpected behavior, especially if the values are different.  On the other hand it doesn't make sense to over optimize when there's no immediate need.
Thoughts?

bjtitus replied:
I think this is definitely a good point for the true/false values. I'm not quite sure about other values since some server implementations will combine duplicates into an array (although, I think PHP takes the last value).

I'd like to submit a follow up with unit tests for this and could add some details and a possible change there. Adding the deduplication could probably prevent some unintended side effects.

Do you think that makes sense?

aerych replied:
Sounds good to me!

aerych commented on WordPress/Classes/ViewRelated/What's New/Data store/StoriesIntroDataSource.swift:
```diff
@@ -0,0 +1,70 @@
+        private let reuseIdentifier = "announcementCell"
+
+        func registerCells(in tableView: UITableView) {
+            tableView.register(Cell.self, forCellReuseIdentifier: reuseIdentifier)
```
Nice :)

aerych commented on WordPress/Classes/ViewRelated/What's New/Data store/StoriesIntroDataSource.swift:
```diff
@@ -0,0 +1,70 @@
+        }
+    }
+
+    var dataDidChange: (() -> Void)?
```
Maybe group this guy with the items prop above?

aerych commented on WordPress/Classes/ViewRelated/What's New/Views/GridCell.swift:
```diff
@@ -0,0 +1,228 @@
+     │└─────────────┘│
+     │ mainStackView │
+     └───────────────┘
+
```
Diagrams! 😮, 👍

aerych commented on WordPress/Classes/ViewRelated/What's New/Views/GridCell.swift:
```diff
@@ -0,0 +1,228 @@
+
+ See `makeItemStackView` for a description of `item`.
+*/
+class GridCell: UITableViewCell {
```
Curious if there was an advantage to composing the UI via code vs a xib?

bjtitus replied:
I mostly just did this to match the `AnnouncementCell`. I don't have a strong preference here but I think we've been moving towards defining relatively small views in code.

aerych reviewed (approved):
Howdy @bjtitus 👋 
Looking good to me!  I just had a couple of nitpicky questions code wise. 
While testing I checked R2L languages and that looked sharp.  I also tried landscape orientation on the iPhone and I wanted to double check if the first line of text was supposed to be left aligned or if the positioning is expected:
![Simulator Screen Shot - iPhone 11 Pro - 2020-10-19 at 10 44 01](https://user-images.githubusercontent.com/1435271/96474937-00083980-11f9-11eb-86c5-b7870e2496ae.png)

:shipit: from me assuming the alignment looks good and the other questions don't spark bigger changes. :)

jkmassel commented:
👋 The `16.0` code freeze is happening today, so I'm bumping this to `16.1`. If it needs to be in `16.0`, please merge it into the `release/16.0` branch and DM me – I'll be happy to cut a new beta!

bjtitus commented:
@megsfulton Do you think devices with compact widths (the break seems to be around 820 pixels) need to switch to left alignment? You can see a full list of size classes with devices here: https://developer.apple.com/design/human-interface-guidelines/ios/visual-design/adaptivity-and-layout#device-size-classes

I'd prefer to keep the switch grounded in Apple's provided size classes since it makes things a bit simpler to implement and maintain and keeps consistency with the platform.

#### Center Aligned
<a href="https://user-images.githubusercontent.com/3250/96619392-5e015380-12c3-11eb-8c19-2abb7d55a7de.png"><img src="https://user-images.githubusercontent.com/3250/96619392-5e015380-12c3-11eb-8c19-2abb7d55a7de.png" width="400"></a>

#### Left Aligned
<a href="https://user-images.githubusercontent.com/3250/96620546-d0266800-12c4-11eb-9b69-d1696714a842.png"><img src="https://user-images.githubusercontent.com/3250/96620546-d0266800-12c4-11eb-9b69-d1696714a842.png" width="400"></a>

I believe the iPhone 11 Pro is the widest Compact Width screen (at 812px):
<a href="https://user-images.githubusercontent.com/3250/96622606-65c2f700-12c7-11eb-8a68-c5b4c1b20ce8.png"><img src="https://user-images.githubusercontent.com/3250/96622606-65c2f700-12c7-11eb-8a68-c5b4c1b20ce8.png" width="400"></a>

megsfulton commented:
I think I'd like to left align everything all the time on this screen. So the example stories are always left aligned with the text.

I'd also like to remove the example story titles below the stories.

bjtitus committed:
Stories Intro changes from design + PR feedback

- Removes the labels from the example stories
- Always left align the example stories
- Moves some code based on suggestions

bjtitus commented:
@aerych I was hoping you could do a quick verification since we had a few design changes to mirror some Android feedback: https://github.com/Automattic/stories-android/issues/575

- The example story titles were removed
- Stories are always left aligned instead of centered (should address your comment about the devices which are regular width in landscape that you mentioned in your previous review).

I updated the screenshots so you should be able to see what it's supposed to look like.

aerych reviewed (approved):
Looks good to me! :shipit: sir!

Gio2018 commented on WordPress/Classes/ViewRelated/What's New/Views/WhatIsNewViewController.swift:
```diff
@@ -6,17 +6,26 @@ class WhatIsNewViewController: UIViewController {
     private let makeWhatIsNewView: () -> WhatIsNewView
 
+    private let onContinue: () -> Void
+    private let onDismiss: (() -> Void)?
```
Sorry I'm late: should we make these two the same type, or is there a reason why one is optional and the other isn't?

Gio2018 reviewed (commented):
Sorry for the delay, this looks great @bjtitus ! I only had one small comment
